### PR TITLE
Keep p5.strands transpiler output valid for comma-joined expressions

### LINKED ISSUE #9178: [p5.js 2.0+ Bug Report]: strands not handling comma operator (common in minified js)
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] WebGPU
- [x] p5.strands
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### p5.js version

2.3.2

### Web browser and version

All

### Operating system

All

### Steps to reproduce this

A common thing you'll see in some JavaScript minifiers is to collapse multiple *statements* (things separated by semicolons generally) into one *expression* (something that evaluates to a single value) via the comma operator. E.g. something like this:
```js
let a
if (something) {
  someThingIsTrue();
  a = 1;
} else {
  somethingElse();
  a = 2;
}
```
...might get compressed into this:
```js
let a = something ? (somethingIsTrue(), 1) : (somethingElse(), 2);
```

Anyway it seems p5.strands does not handle comma operators well. In the example below, simulating something a minifier produced while using strands at work, transpiles to invalid javascript:

```js
let gradientShader

function setup() {
  createCanvas(400, 400, WEBGL)
  gradientShader = buildMaterialShader(() => {
    let pos = sharedVec4()

    worldInputs.begin(), pos = [worldInputs.position/200, 1], worldInputs.end()
    
    pixelInputs.begin(), pixelInputs.color = pos, pixelInputs.end()
  })
}

function draw() {
  clear()
  shader(gradientShader)
  noStroke()
  plane(width, height)
}
```

Live: https://editor.p5js.org/davepagurek/sketches/zNjhdJCGO

This transpiles to:
```js
let pos = sharedVec4('pos');
    worldInputs.begin(), pos.bridge(__p5.strandsNode([
        __p5.strandsNode(worldInputs.position).div(200),
        1
    ]));, worldInputs.end();
    pixelInputs.begin(), pixelInputs.color = pos.getValue(), pixelInputs.end();
```

Note that there's a `;,` in there. I think a few spots in our transpiler code have the assumption that each line is its own statement and is emitting semicolons where it shouldn't. We should do a little audit for those and create some test cases and fixes.

### PR DESCRIPTION
**Pull request by harshiltewari2004**
Resolves #9178

### Changes

The transpiler assumed each call or assignment is its own statement. Comma-joined
code (as minifiers produce) breaks that assumption in three places, all in
`src/strands/strands_transpiler.js`:

**1. Assignments were rewritten into statements.** `AssignmentExpression` turned
the node into an `ExpressionStatement` for the `bridge` and `bridgeSwizzle`
paths, so escodegen emitted a `;` mid-expression (`;,`) and the callback failed
to build. They are now rewritten as call expressions in place, via a new
`replaceWithMethodCall()` helper that the existing computed-member branch also
uses. This covers ternaries, call arguments and `x++` (which routes through the
same handler). In ordinary statement position it also removes a stray empty
statement that was previously emitted.

**2. `.set()` in control flow dropped sibling expressions.** The rewrite replaced
the whole enclosing statement, so in `a(), hook.set(v), b()` the calls to `a()`
and `b()` disappeared from the output with no error. Only the `.set()` call is
replaced now.

**3. `.begin()` / `.end()` were only found as standalone statements.** With
`hook.end(), b();` the generated `hook.set(...)` was inserted after `.end()`,
which produces a hook function that never assigns its result — a GLSL compile
error (`'HOOK_getColor': Function does not return a value`). A new
`statementCallsHookMethod()` helper, shared by both lookups, also looks inside
comma expressions.

### Tests

Four tests in a new `comma operator (#9178)` suite in
`test/unit/webgl/p5.Shader.js`, one per failure mode. Verified by stashing the
source change: all four fail without it, each for its own reason, and pass with
it. The full WebGL and WebGPU shader suites pass.

### PR Checklist

- [x] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [x] [Unit tests] are included / updated

### CODE REVIEW

**Review by Ayush4958 — APPROVED** [review]
LGTM, In place AST mutations correctly preserve minified SequenceExpression struct without dropping sibling node & acorn walk post order traversal ensures the node property deletions are entirely safe.

cc: @davepagurek

File: test/unit/webgl/p5.Shader.js [hunk 1]
```diff
@@ -3357,6 +3357,110 @@ suite('p5.Shader', function () {
       assert.approximately(pixelColor[2], 0, 5);
     });
 
+        suite('comma operator (#9178)', () => {
```
Comment (davepagurek) [inline test/unit/webgl/p5.Shader.js]: Pretty minor but it looks like indentation is off here

File: src/strands/strands_transpiler.js [hunk 2]
```diff
@@ -1697,25 +1693,16 @@ function transformFunctionSetCalls(functionNode) {
...
-              // Replace the .set() call with an assignment
-              parentStmt.type = 'ExpressionStatement';
-              parentStmt.expression = {
-                type: 'AssignmentExpression',
-                operator: '=',
-                left: { type: 'Identifier', name: intermediateVarName },
-                right: node.arguments[0]
-              };
-            }
+             // Replace the .set() call itself with an assignment, in place.
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js]: minor, there's an extra space here

File: src/strands/strands_transpiler.js [hunk 3]
```diff
@@ -1662,20 +1667,11 @@ function transformFunctionSetCalls(functionNode) {
...
+        statementCallsHookMethod(functionNode.body.body[i], 'begin', exprString)
       ) {
-        const beginExprString = escodegen.generate(
-          stmt.expression.callee.object
-        );
-        if (beginExprString === exprString) {
-          beginCallIndex = i;
-          break;
-        }
+        beginCallIndex = i;
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js:1673]: I believe this is still finding the index of the *statement* that contains a hook `.begin()`. If we are trying to splice in `intermediateVarDecl`, so I was wondering if there is a chance that now we're splicing it after it gets used if it happens to get used in the same statement in a different comma-separated expression. Looks like this was only added to handle loops and branches (see https://github.com/processing/p5.js/pull/8576), but does this run on every `set` call? e.g. would this affect something that calls `set` right after a `begin` in a comma-separated expression currently? If not I'm curious what the flow is, would be good to add in a comment to help future contributors understand.

**Review by davepagurek — COMMENTED** [review]
The approach looks good! I left a comment about what looks like it might be an edge case (but maybe not, worth testing to see), let me know what you think!

File: test/unit/webgl/p5.Shader.js (hunk 1 above)
Comment (harshiltewari2004) [reply test/unit/webgl/p5.Shader.js]: Fixed, thankss!

File: src/strands/strands_transpiler.js (hunk 2 above)
Comment (harshiltewari2004) [reply src/strands/strands_transpiler.js]: Fixed.

File: src/strands/strands_transpiler.js (hunk 3 above)
Comment (harshiltewari2004) [reply src/strands/strands_transpiler.js:1673]: Good catch, this was a real bug. With `hook.begin(), hook.set(x)` in one
statement, the `.set()` becomes an assignment to the intermediate variable,
but the `let` was inserted after that statement, so it threw a
ReferenceError. Before this PR it didn't happen only because `.begin()`
wasn't found inside comma expressions, so the declaration went to the top.

On your question: the pass only runs for functions with a `.set()` inside
if/for, but once it runs it rewrites every `.set()` on that hook, including
one next to `.begin()`.

I now insert the declaration before the `.begin()` statement instead of
after, added a comment explaining the flow, and added a regression test for
this case.

File: src/strands/strands_transpiler.js [hunk 4]
```diff
@@ -442,6 +442,29 @@ function addCopyingAndReturn(functionBody, varsToReturn, sourcePrefix = null) {
   });
 }
 
+/**
```
Comment (davepagurek) [inline src/strands/strands_transpiler.js]: Minor but using `**` instead of `*` turns this into a documentation comment, which shows up in the p5 reference. For internal functions, we should just use `*` to make a regular block comment.

File: src/strands/strands_transpiler.js (hunk 3 above)
Comment (davepagurek) [reply src/strands/strands_transpiler.js:1673]: Would there be issues if you had multiple cooks all in the same statement, chained with commas? would it be more robust to see if it's in a comma expression or not, splice it into that right before the begin if so, and splice a statement before the whole statement if not?

File: src/strands/strands_transpiler.js (hunk 4 above)
Comment (harshiltewari2004) [reply src/strands/strands_transpiler.js]: Fixed — changed both new helpers to a regular block comment.

**Comment by harshiltewari2004** [comment]
I tested two hooks sharing one comma statement:

```js
hookA.begin(), hookB.begin();
if (c) { hookA.set([1, 0, 0, 1]); hookB.set([0, 1, 0, 1]); }
hookA.end(), hookB.end();
```

Both declarations land before the shared statement and each `.set()` still
comes before its `.end()`, so I didn't find a problem. Each hook is handled
in its own pass over the statement list, so the shifted indices stay correct.

One thing on splicing into the comma expression: `let __hook_value;` is a
statement, so it can't be an element of a sequence expression. The options
I see are what I have now (always insert before the statement) or a
conditional version that only does that for comma expressions and keeps
insert-after-begin otherwise, which would leave existing output unchanged.
Happy to switch to the conditional one if you'd prefer that.

**Comment by davepagurek** [comment]
Right, I think the approach you have now seems good then. thanks for looking into it!

**Review by davepagurek — APPROVED** [review]
